### PR TITLE
feat: add external source identifier config model

### PR DESCRIPTION
**Pull request by ansabgillani**
[PROD-3086](https://2u-internal.atlassian.net/browse/PROD-3086)

This PR aims to track where the relevant data residing in Discovery is coming from. We have introduced a config model similar to Organizations and Partner to track source (edX or other partners) for each course and program. When implemented, these sources can be populated from Publisher and/or via ingestion in Discovery API.

### CODE REVIEW

File: course_discovery/apps/course_metadata/models.py [hunk 1]
```diff
@@ -96,6 +96,20 @@ class Meta:
         abstract = True
 
 
+class ExternalSourceIdentifier(TimeStampedModel):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: Rename pending, either ProductSource or Source

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -96,6 +96,20 @@ class Meta:
...
+    COURSES = 'course'
+    PROGRAMS = 'program'
+    ITEM_TYPE = (
+        (COURSES, _('Courses')),
+        (PROGRAMS, _('Programs')),
+    )
+
+    name = models.CharField(max_length=255, help_text=_('Name of the external source.'))
+    slug = models.CharField(max_length=255, help_text=_('Slug of the external source identifier.'))
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: Course and Program use auto-generated slug from title/name. Use the similar approach here too.

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -96,6 +96,20 @@ class Meta:
...
+    description = models.CharField(max_length=255, blank=True, help_text=_('Description of the external source.'))
+    type = models.CharField(max_length=255, blank=True, help_text=_('Type of the external source identifier.'))
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: do we really need it? We can use reverse relation on Course/Program to get this info. Maybe I am missing some context here.

File: course_discovery/apps/course_metadata/models.py [hunk 4]
```diff
@@ -1082,6 +1096,7 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
     enrollment_count = models.IntegerField(
         null=True, blank=True, default=0, help_text=_('Total number of learners who have enrolled in this course')
     )
+    external_identifier = models.ForeignKey(ExternalSourceIdentifier, models.SET_NULL, null=True, blank=True)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: add related_name to allow for easy reverse lookup

File: course_discovery/apps/api/serializers.py [hunk 5]
```diff
@@ -2109,8 +2134,26 @@ class Meta(MinimalProgramSerializer.Meta):
...
+            instance.external_identifier = ExternalSourceIdentifier.objects.create(
+                **external_identifier,
+                type='program'
+            )
+            instance.save()
+
+    def update(self, instance, validated_data):
+        if 'external_identifier' in validated_data:
+            self.update_external_identifier(instance, validated_data.pop('external_identifier'))
+        return super().update(instance, validated_data)
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: FYI, Program API currently does not support the creation of objects via API.

**Review by DawoudSheraz — COMMENTED** [review]
Left a few comments.

File: course_discovery/apps/course_metadata/models.py [hunk 6]
```diff
@@ -96,6 +96,18 @@ class Meta:
         abstract = True
 
 
+class Source(TimeStampedModel):
+    """
+    Source Model to find where a course or program originated
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: nit: originated from.

File: course_discovery/apps/course_metadata/models.py [hunk 7]
```diff
@@ -1082,6 +1094,7 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
     enrollment_count = models.IntegerField(
         null=True, blank=True, default=0, help_text=_('Total number of learners who have enrolled in this course')
     )
+    external_product_source = models.ForeignKey(Source, models.SET_NULL, null=True, blank=True, related_name='courses')
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: nit: product_source should suffice.

File: course_discovery/apps/api/serializers.py [hunk 8]
```diff
@@ -225,6 +225,13 @@ class Meta(TitleDescriptionSerializer.Meta):
         model = AdditionalPromoArea
 
 
+class SourceSerializer(BaseModelSerializer):
+    """Serializer for Source"""
+    class Meta:
+        model = Source
+        fields = ('name', 'slug')
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: shouldn't description be added here too?

**Review by DawoudSheraz — APPROVED** [review]
A few nits/qs, otherwise 👍🏽

File: course_discovery/apps/api/serializers.py [hunk 9]
```diff
@@ -2048,6 +2072,7 @@ class ProgramSerializer(MinimalProgramSerializer):
     in_year_value = ProductValueSerializer(required=False)
     skill_names = serializers.SerializerMethodField()
     skills = serializers.SerializerMethodField()
+    external_product_source = SourceSerializer(required=False)
```
Comment (AliAdnanSohail) [inline course_discovery/apps/api/serializers.py]: nit: remove `external`, product_source is enough. As this is not external source this is our source

**Review by AliAdnanSohail — APPROVED** [review]

File: course_discovery/apps/api/serializers.py (hunk 9 above)
Comment (ansabgillani) [reply course_discovery/apps/api/serializers.py]: Fixed

File: course_discovery/apps/course_metadata/models.py (hunk 7 above)
Comment (ansabgillani) [reply course_discovery/apps/course_metadata/models.py]: Fixed

File: course_discovery/apps/course_metadata/models.py (hunk 6 above)
Comment (ansabgillani) [reply course_discovery/apps/course_metadata/models.py]: Fixed

File: course_discovery/apps/course_metadata/models.py (hunk 4 above)
Comment (ansabgillani) [reply course_discovery/apps/course_metadata/models.py]: Fixed

File: course_discovery/apps/course_metadata/models.py (hunk 3 above)
Comment (ansabgillani) [reply course_discovery/apps/course_metadata/models.py]: You're right, we can resolve it with related_name. Fixed

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (ansabgillani) [reply course_discovery/apps/course_metadata/models.py]: Fixed

File: course_discovery/apps/course_metadata/models.py (hunk 1 above)
Comment (ansabgillani) [reply course_discovery/apps/course_metadata/models.py]: Fixed

**Pull request merged**